src/openms/source/CHEMISTRY/RibonucleotideDB.cpp

```diff
@@ -43,21 +43,34 @@ using namespace std;
 
 namespace OpenMS
 {
-  RibonucleotideDB::RibonucleotideDB():
-    max_code_length_(0)
+
+  RibonucleotideDB* RibonucleotideDB::getInstance()
   {
-    readFromFile_("CHEMISTRY/Modomics.tsv");
-    readFromFile_("CHEMISTRY/Custom_RNA_modifications.tsv");
+    static RibonucleotideDB* db_ = RibonucleotideDB::initializeRibonucleotideDB();
+    return db_;
   }
 
-  RibonucleotideDB* RibonucleotideDB::getInstance()
+  RibonucleotideDB* RibonucleotideDB::initializeRibonucleotideDB(OpenMS::String modomics_file, OpenMS::String custom_mods_file, OpenMS::String proprietary_mods_file)
+  {
+    static RibonucleotideDB* db_ = new RibonucleotideDB(modomics_file, custom_mods_file, proprietary_mods_file);
```

Comment:
this is not threadsafe.
i.e. multiple threads could still create a new object, but only one will succeed.
You either need https://en.cppreference.com/w/cpp/thread/call_once or make it a 
`static RibonucleotideDB db_(modomics_file, custom_mods_file, proprietary_mods_file);`

You also need to hide the copy and assignment operators (which are defaulted and broken at the moment). Otherwise a user can copy the object obtained using getInstance().

Reply:
having a static RibonucleotideDB is better but 
wouldn't it also be fine to move get rid of the initialize call and directly do the new in getInstance() ?  
https://stackoverflow.com/questions/70643880/leaky-meyers-singleton-is-it-threadsafe

Reply:
so we need to do something like:
https://www.modernescpp.com/index.php/thread-safe-initialization-of-data
```
class MySingleton{

  private:
    static std::once_flag initInstanceFlag;
    static MySingleton* instance;
    MySingleton()= default;
    ~MySingleton()= default;

  public:
    MySingleton(const MySingleton&)= delete;
    MySingleton& operator=(const MySingleton&)= delete;

    static MySingleton* getInstance(){
      std::call_once(initInstanceFlag,MySingleton::initSingleton);
      return instance;
    }

    static void initSingleton(){
      instance= new MySingleton();
    }
};

MySingleton* MySingleton::instance= nullptr;
std::once_flag MySingleton::initInstanceFlag;
```

Reply:
or easier: don't use the pointer and static initialize

Reply:
> and this would also be wrong? https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/CHEMISTRY/ModificationsDB.cpp#L81-L97

yes. same problem

Reply:
> so we need to do something like: https://www.modernescpp.com/index.php/thread-safe-initialization-of-data
> 
> ```
> class MySingleton{
> 
>   private:
>     static std::once_flag initInstanceFlag;
>     static MySingleton* instance;
>     MySingleton()= default;
>     ~MySingleton()= default;
> 
>   public:
>     MySingleton(const MySingleton&)= delete;
>     MySingleton& operator=(const MySingleton&)= delete;
> 
>     static MySingleton* getInstance(){
>       std::call_once(initInstanceFlag,MySingleton::initSingleton);
>       return instance;
>     }
> 
>     static void initSingleton(){
>       instance= new MySingleton();
>     }
> };
> 
> MySingleton* MySingleton::instance= nullptr;
> std::once_flag MySingleton::initInstanceFlag;
> ```

why is 'static void initSingleton()' publicly accessible? That looks wrong...
just ditch the pointers and do
```
static const MySingleton& getInstance(){
      static MySingleton instance(a,b,c);
      return instance;
    }
```

Reply:
What's the advantage of ditching the pointers? We have a couple of `*DB` classes, so I would want to keep the semantics between them similar.

Reply:
they are also not threadsafe.
But if you really want a pointer, then just do
```
static const MySingleton* getInstance(){
      static MySingleton instance(a,b,c);
      return *&instance;
    }
```
to get the old interface and thread-safety